src/common/transformations/src/transformations/common_optimizations/ric_fusion.cpp

```diff
@@ -0,0 +1,607 @@
+    bool can_be_merged_with(const Attribute & other) {
+        return m_order == other.m_order && m_axis == other.m_axis;
+    }
+
+    // When merging two and more attrs for further propagation we have to keep can_be_fused references
+    // for cases when fusion is not possible, so we can update all related attrs.
+    void merge_with(const Attribute & other) {
+        m_can_be_fused.insert(m_can_be_fused.end(),
+                              other.m_can_be_fused.begin(),
+                              other.m_can_be_fused.end());
+    }
+
+    const std::vector<int64_t> & get_order() const { return m_order; }
+
+    void set_order(const std::vector<int64_t> & order) { m_order = order; }
+
+    int64_t get_axis() const { return m_axis; }
+
+    void set_axis(int64_t axis) { m_axis = axis; }
+
+    bool is_final() const { return m_is_final; }
+
+    bool is_initial() const { return m_is_initial; }
+
+private:
+    std::vector<int64_t> m_order;
+    int64_t m_axis;
+
+    // Specifies whether RIC can be fused or not. vector is needed to keep references to other
+    // attributes that were participated during merge.
+    std::vector<std::shared_ptr<bool>> m_can_be_fused;
+
+    // true - means that current RIC attribute is final and can be materialized
+    // false - means that current RIC attribute is temporary and need only for propagation
+    bool m_is_final;
+
+    // true - means that current RIC attribute is an initial attribute and belongs to real RIC output
+    // false - means that current RIC attribute is temporary and need only for propagation
+    bool m_is_initial;
+
+    // Callback specifies the action for RIC materialization for given input port.
+    // In most cases it should insert Gather operation for the input.
+    std::function<void(Input<Node>, const Attribute &)> m_callback =
+            [](Input<Node>, const Attribute &) {};
+};
+
+namespace {
+
+template <typename T>
+using is_port = typename std::enable_if<!std::is_convertible<T, std::shared_ptr<Node>>::value>::type;
+
+template <typename T, typename = is_port<T>>
+void set(T port, const Attribute & ric_attr) {
+    auto & attrs = port.get_rt_info();
+    attrs["reverse_input_channel_index"] = ric_attr;
+}
+
+// Available only for output ports
+void init(Output<Node> output, std::vector<int64_t> order, int64_t axis) {
+    set(output, Attribute(std::move(order), axis, false, true));
+}
+
+template <typename T, typename = is_port<T>>
+bool has(const T & port) {
+    const auto & attrs = port.get_rt_info();
+    return attrs.count("reverse_input_channel_index");
+}
+
+template <typename T, typename = is_port<T>>
+Attribute get(const T & port) {
+    const auto & attrs = port.get_rt_info();
+    auto res = attrs.find("reverse_input_channel_index");
+    if (res != attrs.end()) {
+        return res->second.template as<Attribute>();
+    }
+    throw ngraph_error("reverse_input_channel_index is missing in given port");
```

Comment:
can it be tested somehow?

Reply:
sure

Reply:
It is impossible to simulate this situation...